Rate Books from Search Modal

sk1122 opened the pull request:
## Description
- Copied rateStart(), rateBook(), finishRate() to startRatingModal(), rateBookModal(), finishRatesModal() with minor changes
- bookid can't be used in layout.html, So using a global variable book__id which will update everytime a Modal opens to the id of Book in Modal

## Affected Dependencies
None

## How has this been tested?

## GIF Before

![Before](https://user-images.githubusercontent.com/40713709/111873353-ef313f00-89b5-11eb-9554-2a1322c1ba6c.gif)



## GIF After
![After](https://user-images.githubusercontent.com/40713709/111873356-f1939900-89b5-11eb-82e6-985b34aee5e3.gif)




## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/Praful932/Kitabe/blob/master/CONTRIBUTING.md).
- [x] My changes do not edit/add any unrequired files.
- [x] My changes are covered by tests.

Praful932 reviewed (changes requested):
@sk1122 Please update the PR Description

Praful932 reviewed (changes requested):
🔴Your changes are affecting existing rating button design on the whole application, please make separate design class & ids for rating functionality in the modal.
See the deployed version for more details

Praful932 commented on templates/mainapp/layout.html:
```diff
@@ -125,7 +125,10 @@ <h4 class="modal-title" id="booktitle">Basic Modal</h4>
                             </p>
 
-                            <div id="bookrating"></div>
+                            <div class='rating-block-modal' id='rating-block-{{ book.book_id }}' onmouseover="startRatingModal(this)" 
```
It's unclear to me how exactly is this `{{ book.book_id }}` being set here, there's no object being iterated for you to access this?
The flow initally to get the book details to modal is like this, whenever the user enters, automatically the corresponding book names link with attributes(see this [line](https://github.com/Praful932/Kitabe/blob/1866e1ff9c8a1164365a449f2c886022a07a65b3/templates/mainapp/layout.html#L247)) are created and when clicked the details are passed onto this function [here](https://github.com/Praful932/Kitabe/blob/1866e1ff9c8a1164365a449f2c886022a07a65b3/templates/mainapp/layout.html#L209), I'm unsure how you are trying to capture the book id here.

sk1122 replied:
I forget to delete this, but now it uses `rate-` like classes or id. We use book__id to get id from the book and send an ajax request to API.

Praful932 commented on templates/mainapp/layout.html:
```diff
@@ -212,9 +218,8 @@ <h4 class="modal-title" id="booktitle">Basic Modal</h4>
-            $("#bookrating").rateYo({
-                rating: averagerating,
-                readOnly: true
+            $(".rateYo").rateYo({
```
You cannot set by using the class, are `rateYo` elements will be set to this property if you do this

sk1122 replied:
Forgot to change. Changed in new Commit

Praful932 commented on templates/mainapp/layout.html:
```diff
@@ -125,7 +125,16 @@ <h4 class="modal-title" id="booktitle">Basic Modal</h4>
+                            {% if user.is_authenticated %}
+                            <div class='rating-block-modal' id='rating-block-' onmouseover="startRatingModal(this)" 
+                            onclick='rateBookModal(this)' onmouseout='finishRatesModal(this)'>
+                                <div id='rateYo' class="rate-" style="display: block"></div>
```
Would suggest to change the id to something like `rateYo-modal`

Praful932 commented on templates/mainapp/layout.html:
```diff
@@ -125,7 +125,16 @@ <h4 class="modal-title" id="booktitle">Basic Modal</h4>
+                                <div id='rateYo' class="rate-" style="display: block"></div>
+                            </div>
+                            {% else %}
+                            <div class="prompt-login">
```
For consistency, the behaviour when not logged in should be similar to what happens on other pages

sk1122 replied:
You mean that I should add this code (copied from `recommendations.html`)

```
<div class='rating-block btn' onclick="location.href='{% url 'account_signup' %}'">
    <div id='button-text'>Rate <i class="fa fa-heart" style="color:red"></i> This Book</a>
    </div>
</div>
```

Praful932 replied:
Yes

sk1122 replied:
![Result](https://user-images.githubusercontent.com/40713709/112143667-39a4fc80-8bfe-11eb-906b-9f6d10365b01.png)
It is looking something like this

Praful932 replied:
Yes

Praful932 commented on templates/mainapp/layout.html:
```diff
@@ -278,26 +289,53 @@ <h4 class="modal-title" id="booktitle">Basic Modal</h4>
             var bookid = obj.getAttribute('data-book-id');
             var bookrating = $('#rate-' + bookid).rateYo('rating');
-            $('#rating-block-' + bookid).css('transition', 'background-color 2s ease');
-            $('#rating-block-' + bookid).css('background-color', '#29BF12');
```
The rating book behavious should be same as earlier

Praful932 commented:
@sk1122 Also your branch seems behind of latest master

sk1122 commented:
@Praful932 should i run the `git pull upstream` command

Praful932 commented:
> @Praful932 should i run the `git pull upstream` command and then commit

Update your fork, and do a [rebase](https://stackoverflow.com/questions/5340724/get-changes-from-master-into-branch-in-git)

Praful932 reviewed (changes requested):
Please don't mark conversations as resolved, make changes and let us review, then we'll resolve :)

sk1122 commented:
@Praful932 i have resolved them in local system and waiting for your confirmation on 2nd issue, then I will commit my changes.

But I am sorry

Praful932 reviewed (commented):
No problem, it's fine

Praful932 commented on templates/mainapp/layout.html:
```diff
@@ -349,7 +409,7 @@ <h4 class="modal-title" id="booktitle">Basic Modal</h4>
-                    $("#bookrating").rateYo({
+                    $("#rateYo").rateYo({
                         rating: averagerating,
                         readOnly: true
```
If it is read only, how is it that I can rate? Could you clarify?

sk1122 replied:
#rateYo is not the id of Rating Stars in Modal.

It is #rateYo-modal

Praful932 replied:
I understand that it was earlier there for the average rating, this function is for when the modal is opened from `More Details` Button, so you need to change this too, to `#rateYo-modal`

sk1122 replied:
So, i should change that and remove readOnly?

Praful932 replied:
Yes, confirm the behaviour, and let me know when it's ready for review

sk1122 commented on templates/mainapp/layout.html:
```diff
@@ -125,7 +125,17 @@ <h4 class="modal-title" id="booktitle">Basic Modal</h4>
+                            {% if user.is_authenticated %}
+                            <div class='rating-block-modal' id='rating-block-' onmouseover="startRatingModal(this)" 
+                            onclick='rateBookModal(this)' onmouseout='finishRatesModal(this)'>
+                                <div id='rateYo-modal' class="rate-" style="display: block"></div>
```
You can check it right here

sk1122 commented:
![kitabe](https://user-images.githubusercontent.com/40713709/112635172-7e7c9d80-8e61-11eb-9264-84fe0bb0b4b6.gif)

I have not edited any of the code related to rating outside of the element, but still how is that happening?

id of that element is rateYo but in Firefox Dev Tools it is showing class = "rateYo".

I have updated the repo with the latest update.

Please check if its happening in yours also

Praful932 commented:
@sk1122 Because it [is](https://github.com/Praful932/Kitabe/blob/master/templates/mainapp/explore.html#L37)

Praful932 commented:
@sk1122 Re-request a review when it's ready :)

sk1122 commented:
> @sk1122 Re-request a review when it's ready :)

It is working fine, I checked on my local machine

Praful932 reviewed (changes requested):
`book__id` can be removed now..

Praful932 commented:
Also instead of keeping average_rating when user opens modal, make it `0`

sk1122 commented:
`book__id` is required as we have to POST details to server

Praful932 commented on templates/mainapp/layout.html:
```diff
@@ -329,6 +382,7 @@ <h4 class="modal-title" id="booktitle">Basic Modal</h4>
         // Retrieve and display book details in Modal
         function bookDetails(obj) {
             var bookid = obj.getAttribute('data-book-id');
+            book__id = bookid;
```
Are you sure? I'm only seeing it being used here

Praful932 replied:
Any updates?

Praful932 commented:
@sk1122 Any updates?

sk1122 commented:
@Praful932 i have reviewed changes and pointed where book__id is used a week ago...

Please take some time to check it

Praful932 commented:
@sk1122 Is this not visible on your side?
![image](https://user-images.githubusercontent.com/45713796/114305094-46798980-9af4-11eb-8b5a-19d5196c6a40.png)

sk1122 commented:
It is visible, but is this visible on your side, in which i mentioned where it is used

@Praful932 

![Screenshot_2021-04-11-18-34-08-272_com github android](https://user-images.githubusercontent.com/40713709/114305195-a8d28a00-9af4-11eb-8077-ad35d5c51848.png)

Praful932 commented:
@sk1122 You started a review... see the pending icon, means it won't be visible on my side, try visiting the same page when not logged in..
You clicked on **start a review** instead of **add a single comment**..

sk1122 commented:
@Praful932 ok, but you have seen that block of code in SS

Praful932 commented:
Could you delete those and add the ones you sent as ss, It's easier to communicate in the code

Praful932 commented on templates/mainapp/layout.html:
```diff
@@ -278,26 +290,58 @@ <h4 class="modal-title" id="booktitle">Basic Modal</h4>
         }
         // onclick rate
+        function rateBookModal(obj) {
+            var bookid = obj.getAttribute('data-book-id');
```
Could you clear this out? If `bookid` is being captured here, what's the need of using `book__id` in further lines?

sk1122 replied:
`bookid` is an empty string because data-book-id doesn't exist for that block of html

Praful932 reviewed (changes requested):
@sk1122 I have added some comments, could you check?

Praful932 commented on templates/mainapp/layout.html:
```diff
@@ -125,7 +125,17 @@ <h4 class="modal-title" id="booktitle">Basic Modal</h4>
-                            <div id="bookrating"></div>
+                            {% if user.is_authenticated %}
+                            <div class='rating-block-modal' id='rating-block-modal'
+                            onclick='rateBookModal(this)' onmouseout='finishRatesModal(this)'>
```
@sk1122 Instead of making a new global variable, why don't you set a new attribute on this element as `data-book-id` when the modal is modified in the function at Line 383 then you won't have to use the new variable in the function at  Line 318, Does that make sense?

sk1122 replied:
I tried that but how to access the id,

`explore.html` template is being rendered by function and `layout.html` `extends` in explore.html, So we cannot send variables directly to `layout.html` so we cannot set `data-book-id` in layout.html

Praful932 replied:
Set the attribute using setattr or a similar function in js on line 383 when the modal is modified

Praful932 commented:
Code seems fine to me now, Could you add a gif/video showing first in the [admin](https://www.tutorialspoint.com/django/django_admin_interface.htm) panel that the rating does not exist then rating it through search and then it appears in the admin panel as well as library

sk1122 commented:
Demo - 

Uploading kitabedemo.mp4…

sk1122 commented:
Demo


https://user-images.githubusercontent.com/40713709/114876272-ca48b400-9e1b-11eb-970a-2a06ae2b2c1d.mp4

Praful932 commented:
Update your branch with latest master so we can review

sk1122 commented:
It is already up to date

![uptodate](https://user-images.githubusercontent.com/40713709/115001504-52cd6000-9ec1-11eb-8675-7c3f951501f7.png)

Praful932 commented:
@sk1122 Update your fork see [starter guide FAQ](https://github.com/Praful932/Kitabe/discussions/42)

sk1122 commented:
I updated the fork from Upriver but
![kitabeconsole](https://user-images.githubusercontent.com/40713709/115134213-c68b7c00-a02b-11eb-9c88-da697c968236.png)

Praful932 commented:
Seems the earlier commit updated it.
There's some issue with Rate buttons across the application, Check the online deployment
![image](https://user-images.githubusercontent.com/45713796/115138139-05c7c600-a048-11eb-826b-2a49bdf19bb2.png)

sk1122 commented:
@Praful932 It is returning server error on kitabe's website

Praful932 commented:
> @Praful932 It is returning server error on kitabe's website

Seems we are running out of heroku free tier memory, should we working now

sk1122 commented:
@Praful932 can you specify what you wanted to say in the previous comment?

I checked on deployment but the rating button inside the modal is not working as thought.

Praful932 commented:
Hopefully [this](https://www.loom.com/share/b87ff195798941989c507cfee1dbed5a) makes sense,
The first half is this PR & the second is the latest deployment

sk1122 commented:
Got it, will start woking on it!!!

sk1122 commented:
`<div id='rateYo rate-{{ book.book_id }}'>` (line 82, explore.html), I have updated it in local dev but it is not updating it in html

![kitaberating](https://user-images.githubusercontent.com/40713709/115360491-4b69c780-a1dd-11eb-97ea-1bd38cc72a42.png)

Praful932 commented:
Try clearing cached and reload `ctrl + shift +R`  or try in incognito/another browser

sk1122 commented:
> Try clearing cached and reload `ctrl + shift +R`  or try in incognito/another browser

I tried it already

sk1122 commented:
Is it possible that it is loading files from git repository?

Praful932 commented:
> Is it possible that it is loading files from git repository?

Nope, could you try to contact one of the mentors/admins on discord?
We can help you there

sk1122 commented:
Ok

sk1122 commented:
Updated Fork, and everything is working fine.

Praful932 reviewed (changes requested):
Fork seems to be outdated :/ Could you update your fork

sk1122 commented:
I have updated the fork, but I wasn't able to push the code so I added a new line where needed `layout.html` which helped to push the code with an updated fork.

Praful932 commented:
Still seems to be outdated

sk1122 commented:
![kitabegithub](https://user-images.githubusercontent.com/40713709/116558865-82e31d00-a91d-11eb-8a21-ddd9e8b7ba73.png)

It is even with master branch

Praful932 commented:
But [this](https://github.com/sk1122/Kitabe/tree/sk1122-rate-search-37) one is not

sk1122 commented:
I force updated that branch and it merged and automatically closed this issue, how this happened?

Praful932 commented:
Force pushing made it even with master branch and erased your commit! Since now both the branches are same(no changes), it automatically closed your PR
That's why never force push. Do you have the changes in your system, if so make a fresh PR,
if the changes seem to be gone, do `git reflog`
